File: algorithms/python/h-Index/h-Index.py
#! /usr/bin/env python3
# -*- coding: utf-8 -*-

# Source: https://leetcode.com/problems/h-index/
# Author: Miao Zhang
# Date:   2021-01-29

class Solution:
    def hIndex(self, citations: List[int]) -> int:
        n = len(citations)
        citations.sort()
        res = 0
        for i in range(len(citations)):
            res = max(res, min(n - i, citations[i]))
        return res

File: algorithms/cpp/h-Index/h-Index.cpp
// Source: https://leetcode.com/problems/h-index/
// Author: Miao Zhang
// Date:   2021-01-28

class Solution {
public:
    int hIndex(vector<int>& citations) {
        int n = citations.size();
        sort(citations.begin(), citations.end());
        int l = 0;
        int r = n - 1;
        int res = 0;
        while (l <= r) {
            int mid = (l + r) / 2;
            res = max(res, min(citations[mid], n - mid));
            if (citations[mid] < n - mid) {
                l = mid + 1;
            } else {
                r = mid - 1;
            }
        }
        return res;
    }
};
